README.md
# Snake_NN
AI learn to play snale via genetic algorithm


history.txt
2020_07_16_22_53
2020_07_20_20_03
2020_07_22_00_30


test_routines.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jul 23 16:03:13 2020

@author: User
"""
from main import save_best, read_best, save_status, load_status
import os
import datetime
from IA_GPU import IA_GPU


#test read/save best    
def test_save_best():
    now = datetime.datetime.now()
    newDirName = now.strftime("%Y_%m_%d_%H_%M")
    os.mkdir(newDirName)
    ia=[IA_GPU()]*10
    for i in range(10):
        ia[i]=IA_GPU()
        ia[i].fill_layers()
    save_best(ia, newDirName)

def test_read_best():
    ia_best=read_best('2020_07_23_16_08')   
    print(ia_best[0].get_weights())

#test read/save status    
def test_save_status():
    now = datetime.datetime.now()
    newDirName = now.strftime("%Y_%m_%d_%H_%M")
    os.mkdir(newDirName)
    ia=[IA_GPU()]*10
    for i in range(10):
        ia[i]=IA_GPU()
        ia[i].fill_layers()
    save_status(ia, newDirName)


def test_load_status():
    ia=load_status('2020_07_23_16_10')   
    print(ia[0].get_weights())

#test_save_best()
#test_read_best()
#test_save_status()
#test_load_status()

test_routines_GPU.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jul  9 15:01:10 2020

@author: User
"""
from IA_GPU import IA_GPU, mutation, recombination, recombination_2
from datetime import datetime
import sys
import os
import numpy as np
from numba import cuda
from numba.cuda.random import create_xoroshiro128p_states

def test_predict():
    ia=IA_GPU()
    ia.fill_layers()
    arr=ia.weights_layer_2
    inp_arr=np.array([[1.]]*18)
    inp_arr2=np.array([[1.]]*19)
    out_arr=np.array([[0.]*19])
    move=ia.predict(inp_arr)
    print(move[0])

def test_mutation():
    a=np.array([[0.,0.,0.,0.,0.],[0.,0.,0.,0.,0.]])
    A_global_mem = cuda.to_device(a)
    dt = datetime.now()
    rng_states = create_xoroshiro128p_states(10, seed=dt.microsecond)
    mutation[2,5](A_global_mem,2,3,rng_states,0.3)
    A=A_global_mem.copy_to_host()
    print(A)
    
def test_recombination():
    dt = datetime.now()
    tot_ia=5
    rng_states = create_xoroshiro128p_states(1*tot_ia, seed=dt.microsecond)
    inp_weights_local=np.array([[1.,2.,3.,4.,5.,6.,7.,8.,9.],[9.,8.,7.,6.,5.,4.,3.,2.,1.]])
    inp_weights= cuda.to_device(inp_weights_local)
    out_weights = cuda.device_array((tot_ia, 9))
    recombination[1,tot_ia](inp_weights, out_weights, 2, tot_ia, 9, rng_states)
    C = out_weights.copy_to_host()
    print(C)

def test_recombination_2():
    dt = datetime.now()
    tot_ia=5
    rng_states = create_xoroshiro128p_states(644*tot_ia, seed=dt.microsecond)
    inp_weights_local=np.array([[1.,2.,3.,4.,5.,6.,7.,8.,9.],[9.,8.,7.,6.,5.,4.,3.,2.,1.]])
    inp_weights= cuda.to_device(inp_weights_local)
    out_weights = cuda.device_array((tot_ia, 9))
    recombination_2[tot_ia,9](inp_weights, out_weights, 2, tot_ia, 9, rng_states)
    C = out_weights.copy_to_host()
    print(C)
  
#test_predict()
#test_mutation()
#test_recombination()
#test_recombination_2()

IA.py
# -*- coding: utf-8 -*-
"""
Created on Fri May 29 15:00:02 2020

@author: User
"""
from numpy.random import randint
from random import random
from random import gauss
from grid import grid
import tkinter as tk
from tkinter import Canvas
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from keras.models import Sequential
from keras.layers import Dense, Conv1D, Flatten
#18 input neurons
#16 hidden
#8 hidden or 16 hidden
#4 final

class IA:
    
    def __init__(self):
        self.Grid=grid()
        #create model
        self.model = Sequential()
        #add model layers
        self.model.add(Conv1D(16, kernel_size=1, activation='relu', input_shape=(18,1)))
        self.model.add(Conv1D(8, kernel_size=1, activation='relu'))
        self.model.add(Flatten())
        self.model.add(Dense(4, activation='softmax'))
        
    def get_weights_as_nparray(self):
        array=self.model.get_weights()
        i=0
        Weights = np.arange(0, dtype="float32")
    
        for  x in array:    
            if i==0:
                for y in x[0][0]:
                    Weights=np.append(Weights,[y], axis=0)
                
            elif i==2:
                for y in x[0]:
                    for z in y:
                        Weights=np.append(Weights,[z], axis=0)
            elif i == 4:
                for y in x:
                    for z in y:
                        Weights=np.append(Weights,[z], axis=0)
            elif i%2==1:
                for y in x:
                    Weights=np.append(Weights,[y], axis=0)
            i+=1
        
        return Weights
    
    def set_weights_as_nparray(self, weights):
        j=0 #counter of weights
        array=self.model.get_weights()
        
        for i,  x in enumerate(array):    
            if i==0:
                for k, y in enumerate(x[0][0]):
                    array[i][0][0][k]=weights[j]
                    j+=1
            
            elif i==2:
                for l, y in enumerate(x[0]):
                    for m, z in enumerate(y):
                        array[i][0][l][m]=weights[j]
                        j+=1
        
            elif i==4:              
                for k, y in enumerate(x):
                    for l, z in enumerate(y):
                        array[i][k][l]=weights[j]
                        j+=1
         
            elif i%2==1:
                for k, y in enumerate(x):
                    array[i][k]=weights[j]
                    j+=1
         
            
        self.model.set_weights(array)
        

    def get_points(self):
        return self.Grid.get_points()

    def set_points(self, points):
         self.Grid.set_points(points)
    
    def check_direction(self, versor): 
        boolean = True
        body_found= False
        i=1
        distance_body=50
        distance_wall=50
        while boolean:
            pos=[x + i*y for x, y in zip(self.Grid.snake.position[0], versor)]
            if pos[0]<0 or pos[0]>39 or pos[1]<0 or pos[1]>39:
                distance_wall = i-1
                boolean=False
            else:
                if self.Grid.grid [pos[0]] [pos[1]] == 1:
                    if (not body_found):
                        distance_body = i 
                        body_found=True
            i+=1
        return [distance_body, distance_wall]
                
                
                
    def calculate_distances(self):
        """calculate the distance(in x and y) between the head of the snake an the food. 
           Also looks in 8 directions to check if there is a part of the body of the snake or a wall"""
        self.Grid.update_grid()
        distances=[]
        x_dist=self.Grid.food.x-self.Grid.snake.position[0][0]
        y_dist=self.Grid.food.y-self.Grid.snake.position[0][1]
        distances.append(x_dist)
        distances.append(y_dist)
        
        #versor (1,0)
        distances+=self.check_direction([1,0])
        #versor (1,1)
        distances+=self.check_direction([1,1])
        #versor (0,1)
        distances+=self.check_direction([0,1])
        #versor (-1,1)
        distances+=self.check_direction([-1,1])
        #versor (-1,0)
        distances+=self.check_direction([-1,0])
        #versor (-1,-1)
        distances+=self.check_direction([-1,-1])
        #versor (0,-1)
        distances+=self.check_direction([0,-1])
        #versor (1,-1)
        distances+=self.check_direction([1,-1])
        
        distances_np=np.array(distances)
        return distances_np
    
    def draw(self):
        window=tk.Tk()
        window.geometry("839x839")
        window.title("prova")
        window.resizable(False,False)
        window.configure(background="black")

        c=Canvas(window,height=839,width=839,bg="black")
        c.pack()
    
        self.Grid.update_grid()
    
        squares =[[c.create_rectangle(0,0,20,20,fill="black")]*self.Grid.column for _ in [c.create_rectangle(0,0,20,20,fill="black")]*self.Grid.rows]
        for i, x in enumerate(squares):
            for j, y in enumerate(x):
                squares[i][j]=c.create_rectangle(0+21*i,0+21*j,20+21*i,20+21*j,fill="white")
        
        print (self.Grid.food.x)
        print (self.Grid.food.y)
        print (self.Grid.snake.position[0])
        self.Grid.draw(c, squares)
        
        def keypress(event, ia, c ,squares,window):
            ia.Grid
            if event.char == "a": ia.Grid.move_left()
            elif event.char == "d": ia.Grid.move_right()
            elif event.char == "w": ia.Grid.move_up()
            elif event.char == "s": ia.Grid.move_down()
    
            if ia.Grid.status==False:
                window.destroy()
       
            else:
                ia.Grid.update_grid()
                ia.Grid.draw(c, squares)
                
        window.bind("<Key>", lambda event: keypress(event, self, c, squares,window))
        window.mainloop()

        
    def predict(self, data):
        data=data.reshape(1,18,1)
        return self.model.predict([data])

    def move_up(self):
        self.Grid.move_up()
      
    def move_down(self):
        self.Grid.move_down()
        
    def move_left(self):
        self.Grid.move_left()

    def move_right(self):
        self.Grid.move_right()

    def get_food_pos(self):
        return self.Grid.get_food_pos()
    
    def draw_2(self, canvas, squares):
        self.Grid.draw(canvas,squares)






def get_points(Ia):
    return Ia.get_points()

def order_IA(ia):
    ia.sort(key=get_points, reverse=True)
    return ia

def recombination_1(array_1, array_2, cut):
    a1=array_1.copy()
    a2=array_2.copy()
    for i in range(cut):
        a1[i]=array_2[i]
        a2[i]=array_1[i]
    result=[a1,a2]
    return result
    
def recombination_2(weights):
    result = []
    pos = randint(0, len(weights),len(weights[0]))
    k=0
    
    for i, x in enumerate(pos):
        #print(weights[x][i])
        result.append(weights[x][i])
        k+=1
        
    return result
    
def mutation(weights, mutation_prob=0.01):
    """generate a mutation with a probability of 1% and a mutation that is gaussian with mean 0 and sigma 1"""
    
    array=weights.copy()

    
    for i, x in enumerate(array):
        rand_float=random()
        if rand_float < mutation_prob:
            array[i]+=gauss(0,1)
    return array
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    

grid.py
# -*- coding: utf-8 -*-
"""
Created on Mon May 25 15:13:57 2020

@author: User
"""
from numpy.random import randint
from snake import snake
from food import food
#import tkinter as tk
#0 vuota, 1 serpente, 2 cibo

class grid(object):
    def __init__(self, _rows=40, _column=40,snake_len=4): 
        self.grid = [[0]*_column for _ in [0]*_rows]
        self.rows = _rows
        self.column = _column
        self.snake=snake(Lenght=snake_len)
        self.food=food(rows=self.rows, column=self.column)
        self.status=True
        self.points=0
        self.moves_left=200
        
    def Print(self):
       for x in self.grid:
           for y in x:
               print (y, end=" ")
           print("\n")


    def import_snake(self, Snake):
        self.snake=Snake
        for i, x in enumerate(self.grid):
            for j, y in enumerate(x):
                if y==1:
                    self.grid[j][i]=0
                    
        for x in Snake.position:
            self.grid[x[1]][x[0]]=1


    def draw (self, canvas, squares):
         for i, x in enumerate(self.grid):
            for j, y in enumerate(x):
                if "white"== canvas.itemcget(squares[i][j], "fill"):
                    canvas.itemconfig(squares[i][j],fill="black")
                if y==1:
                    canvas.itemconfig(squares[i][j],fill="white") 
                if y==2:
                    canvas.itemconfig(squares[i][j],fill="red")
                    
                    
    def import_food(self, food):
        if self.grid[food.y][food.x]==1:
            food.set_position(randint(0, self.column, 2))
            self.import_food(food)
        else:
            self.food=food
            self.grid[food.y][food.x]=2
    
    def update_grid (self):  
        for i, x in enumerate(self.grid):
            for j, y in enumerate(x):
                if y==1:
                    self.grid[i][j]=0
                    
        for x in self.snake.position:
            self.grid[x[1]][x[0]]=1            
            
        while self.grid[self.food.y][self.food.x]==1:
            self.food.set_position(randint(0, self.column, 2))
        
        self.grid[self.food.y][self.food.x]=2    

    def collision(self):
        if self.snake.position[0][0]<0 or self.snake.position[0][0]>=self.column or self.snake.position[0][1]<0 or self.snake.position[0][1]>=self.rows:
            self.status=False
     
        for a in self.snake.position[1:]:
            if a==self.snake.position[0]:
                self.status=False    
            
    def eat(self):
        if self.food.x == self.snake.position[0][0] and self.food.y == self.snake.position[0][1]:
            self.snake.lenght+=1
            self.points+=1
            self.moves_left+=100
            self.snake.position.append(self.snake.last_pos)
            self.food.set_position(randint(0, self.column, 2))
           
    def move_up(self):#ok
        if self.moves_left<=0:
            self.status=False
        else:
            self.snake.move_up()
            self.moves_left-=1
            self.collision()
            self.eat()
            if self.status == True:
                self.update_grid()
                
    def move_down(self):
        if self.moves_left<=0:
            self.status=False
        else:
            self.snake.move_down()
            self.moves_left-=1
            self.collision()
            self.eat()
            if self.status == True:
                self.update_grid()
                
    def move_left(self):
        if self.moves_left<=0:
            self.status=False
        else:
            self.snake.move_left()
            self.moves_left-=1
            self.collision()
            self.eat()
            if self.status == True:
                self.update_grid()
        
    def move_right(self):
        if self.moves_left<=0:
            self.status=False
        else:
            self.snake.move_right()
            self.moves_left-=1
            self.collision()
            self.eat()
            if self.status == True:
                self.update_grid()
     
    def get_points(self):
        return self.points            
            
    def set_points(self, points):
        self.points = points
            
    def get_food_pos(self):     
        return self.food.get_position()

    def set_food_pos(self, pos):
        self.food.set_position(pos)

    def restart(self):
        self.status=True
        self.points=0
        self.moves_left=200
        self.snake=snake(4)
        self.food=food(rows=self.rows, column=self.column)
        
    #Functions created for replay the IA, no random generation of the food
    def Eat(self):
        if self.food.x == self.snake.position[0][0] and self.food.y == self.snake.position[0][1]:
            self.snake.lenght+=1
            self.points+=1
            self.moves_left+=100
            self.snake.position.append(self.snake.last_pos)    
        
    def Move_up(self):
        if self.moves_left<=0:
            self.status=False
        else:
            self.snake.move_up()
            self.moves_left-=1
            self.collision()
            self.Eat()
                
    def Move_down(self):
        if self.moves_left<=0:
            self.status=False
        else:
            self.snake.move_down()
            self.moves_left-=1
            self.collision()
            self.Eat()
                
    def Move_left(self):
        if self.moves_left<=0:
            self.status=False
        else:
            self.snake.move_left()
            self.moves_left-=1
            self.collision()
            self.Eat()
        
    def Move_right(self):
        if self.moves_left<=0:
            self.status=False
        else:
            self.snake.move_right()
            self.moves_left-=1
            self.collision()
            self.Eat()  
            
    def Update_grid (self):  
        for i, x in enumerate(self.grid):
            for j, y in enumerate(x):
                if y==1:
                    self.grid[i][j]=0
                    
        for x in self.snake.position:
            self.grid[x[1]][x[0]]=1            
        
        self.grid[self.food.y][self.food.x]=2 

snake.py
# -*- coding: utf-8 -*-
"""
Created on Mon May 25 15:14:20 2020

@author: User
"""
class snake(object):
    def __init__(self, Lenght=4):    
        self.lenght=Lenght
        self.position=[[10,10]]
        self.last_pos=[-1,-1] #indicates the last position of the tail
        for i in range(0,Lenght-1):
            self.position.append([9-i,10])
   
    def move_down(self):  
        prev_position=(self.position[0]).copy()
        self.last_pos=self.position[self.lenght-1]
        i=0
        for x in self.position:
            if i==0:
                self.position[i][0]+=1
            else:
                buff=self.position[i]
                self.position[i]=prev_position
                prev_position=buff
            i+=1
                        
    def move_up(self):
        prev_position=(self.position[0]).copy()
        self.last_pos=self.position[self.lenght-1]
        
        i=0
        for x in self.position:
            if i==0:
                self.position[i][0]-=1
            else:
                buff=self.position[i]
                self.position[i]=prev_position
                prev_position=buff
            i+=1
   
    def move_right(self):
        prev_position=(self.position[0]).copy()
        self.last_pos=self.position[self.lenght-1]
        i=0
        for x in self.position:
            if i==0:
                self.position[i][1]+=1
            else:
                buff=self.position[i]
                self.position[i]=prev_position
                prev_position=buff
            i+=1
    
    def move_left(self):
        prev_position=(self.position[0]).copy()
        self.last_pos=self.position[self.lenght-1]
        
        i=0
        for x in self.position:
            if i==0:
                self.position[i][1]-=1
            else:
                buff=self.position[i]
                self.position[i]=prev_position
                prev_position=buff
            i+=1


snake_IA.py
# -*- coding: utf-8 -*-
"""
Spyder Editor

This is a temporary script file.
"""

#import tkinter as tk
from grid import grid
from snake import snake
from food import food
import tkinter as tk
from tkinter import Canvas

#griglia 40*40
#lunghezza iniziale serpente 4
#posizione iniziale serpente 20,20

griglia=grid(_rows=40, _column=40)
snake=snake()
        
window=tk.Tk()
window.geometry("839x839")
window.title("prova")
window.resizable(False,False)
window.configure(background="black")

c=Canvas(window,height=839,width=839,bg="black")
c.pack()      
        
squares =[[c.create_rectangle(0,0,20,20,fill="black")]*griglia.column for _ in [c.create_rectangle(0,0,20,20,fill="black")]*griglia.rows]

for i, x in enumerate(squares):
    for j, y in enumerate(x):
        squares[i][j]=c.create_rectangle(0+21*i,0+21*j,20+21*i,20+21*j,fill="white")
        
    
cibo=food(rows=griglia.rows, column=griglia.column)
print (cibo.x)
print (cibo.y)
print (snake.position[0])
griglia.import_food(food=cibo)
griglia.import_snake(Snake=snake)
griglia.draw(c, squares)

def keypress(event, griglia, c ,squares,window):
    
    if event.char == "a": griglia.move_left()
    elif event.char == "d": griglia.move_right()
    elif event.char == "w": griglia.move_up()
    elif event.char == "s": griglia.move_down()
    
    if griglia.status==False:
        window.destroy()
       
    else:
        griglia.update_grid()
        griglia.draw(c, squares)

window.bind("<Key>", lambda event: keypress(event, griglia, c, squares,window))
           
if __name__=="__main__":
    window.mainloop()

#print("senza nulla\n")
#griglia.Print()
#print("\n\n\n")

#griglia.import_snake(Snake=snake)
#print("con serpente\n")
#griglia.Print()
#print("\n\n\n")

#snake.move_left()
#snake.move_left()
#snake.move_left()
#snake.move_left()
#snake.move_down()
#griglia.import_snake(Snake=snake)

#print("con serpente mosso\n")
#griglia.Print()
#print("\n\n\n")



















prova_grafica.py
# -*- coding: utf-8 -*-
"""
Created on Sat May 23 11:16:50 2020

@author: User
"""


import tkinter as tk
from tkinter import Canvas

window=tk.Tk()
window.geometry("900x839")
window.title("prova")
window.resizable(False,False)
window.configure(background="white")

c=Canvas(window,height=839,width=839,bg="black")
c1=Canvas(window,height=839,width=61,bg="red")
c.pack(anchor=tk.W)
#c.pack()
c1.pack(anchor=tk.NE)
#canvas.create_rectangle(x1, y1, x2, y2, **kwargs), 
#with (x1,y1) the coordinates of the top left corner and (x2, y2) those of the bottom right corner


#r1=c.create_rectangle(0,0,20,20,fill="red")
#r2=c.create_rectangle(0,0,40,20,fill="blue")
#r3=c.create_rectangle(0,0,20,40,fill="yellow")
#r4=c.create_rectangle(0,60,20,20,fill="green")
#r5=c.create_rectangle(60,0,20,20,fill="white")

#squares =[[c.create_rectangle(0,0,20,20,fill="white")]*40 for _ in [c.create_rectangle(0,0,20,20,fill="white")]*40]

#for i, x in enumerate(squares):
 #   for j, y in enumerate(x):
  #      squares[i][j]=c.create_rectangle(0+21*i,0+21*j,20+21*i,20+21*j,fill="black")



#c.itemconfig(squares[0][20],fill="white")










def keypress(event):
	x, y = 0, 0
	if event.char == "a": x = -10 
	elif event.char == "d": x = 10
	elif event.char == "w": y = -10
	elif event.char == "s": y = 10
	#c.move(squares[0][0], x, y)

#window.bind("<Key>", keypress)

#first_button=tk.Button(text="Saluta!")
#first_button.grid(row=0,column=0)

if __name__=="__main__":
    window.mainloop()
    
    
    
#window.destroy()

IA_GPU.py
# -*- coding: utf-8 -*-
"""
Created on Sat Jun 13 10:36:19 2020

@author: User
"""

from __future__ import print_function, absolute_import
from grid import grid
from numba import cuda
from numba.cuda.random import create_xoroshiro128p_states, xoroshiro128p_uniform_float32, xoroshiro128p_normal_float32
import numpy as np
from datetime import datetime
import math
from scipy.special import expit

@cuda.jit 
def fill_random(arr, size, rng_states):
    pos = cuda.grid(1)
    if pos < size:
       arr[pos]=xoroshiro128p_uniform_float32(rng_states, cuda.grid(1))*2.-1.

@cuda.jit
def scalar(vector,matrix,result,size):
    pos = cuda.grid(1)
    if pos<size:
        result[pos]=(matrix[pos]*vector).sum()

@cuda.jit
def matmul(A, B, C):
    """Perform matrix multiplication of C = A * B
    """
    row, col = cuda.grid(2)
    if row < C.shape[0] and col < C.shape[1]:
        tmp = 0.
        for k in range(A.shape[1]):
            tmp += A[row, k] * B[k, col]
        C[row, col] = tmp

@cuda.jit 
def recombination(inp_weights, out_weights, n_inp_ia, tot_ia, n_weights, rng_states):#n_inp_ia: number of ia in input, tot_ia: total number of ia to be generated, n_weights: total number of weighta for ia
    
    pos = cuda.grid(1)
    if pos < tot_ia:
        ia_1 = int(xoroshiro128p_uniform_float32(rng_states, cuda.grid(1)) * n_inp_ia)
        ia_2 = int(xoroshiro128p_uniform_float32(rng_states, cuda.grid(1)) * n_inp_ia)
        cut = int(xoroshiro128p_uniform_float32(rng_states, cuda.grid(1)) * n_weights)
        for i in range(n_weights):
            if i < cut:
                out_weights[pos][i]=inp_weights[ia_1][i]
            else:
                out_weights[pos][i]=inp_weights[ia_2][i]

@cuda.jit #n block = tot_ia, threads per block = n_weights
def recombination_2(inp_weights, out_weights, n_inp_ia, tot_ia, n_weights, rng_states):
    # Thread id in a 1D block
    tx = cuda.threadIdx.x
    # Block id in a 1D grid
    ty = cuda.blockIdx.x
    if ty<tot_ia and tx<n_weights:
        ia_rng = int(xoroshiro128p_uniform_float32(rng_states, cuda.grid(1)) * n_inp_ia)
        out_weights[ty][tx] = inp_weights[ia_rng][tx]


@cuda.jit
def mutation(inp_weights, n_ia, n_weights, rng_states, prob=0.1):
    # Thread id in a 1D block
    tx = cuda.threadIdx.x
    # Block id in a 1D grid
    ty = cuda.blockIdx.x
    if ty<n_ia and tx<n_weights:
        a=xoroshiro128p_uniform_float32(rng_states, cuda.grid(1))
        if a<prob:
            inp_weights[ty][tx]+=(xoroshiro128p_normal_float32(rng_states, cuda.grid(1)))/5.



class IA_GPU:
    def __init__(self, nodes_in_layer=[18,16,8,4]):
        self.grid=grid()
        self.layers=4
        self.nodes_in_layer=nodes_in_layer
        self.weights_layer_2=np.array([
                [0.]*(18+1)
            ]*16)
        self.weights_layer_3=np.array([
                [0.]*(16+1)
            ]*16)
        self.weights_layer_4=np.array([
                [0.]*(16+1)
            ]*4)
        self.moves_done=[]
        self.food_pos=[]
        
    def fill_layers(self):
        blocks=1
        
        threads_per_block=19
        dt = datetime.now()
        rng_states = create_xoroshiro128p_states(threads_per_block * blocks, seed=dt.microsecond)
        for i, x in enumerate(self.weights_layer_2):
            fill_random[1,threads_per_block](self.weights_layer_2[i], 19, rng_states)
            
        threads_per_block=17
        dt = datetime.now()
        rng_states = create_xoroshiro128p_states(threads_per_block * blocks, seed=dt.microsecond)
        for i, x in enumerate(self.weights_layer_3):
            fill_random[1,threads_per_block](self.weights_layer_3[i], 17, rng_states)
           
        threads_per_block=17
        dt = datetime.now()
        rng_states = create_xoroshiro128p_states(threads_per_block * blocks, seed=dt.microsecond)
        for i, x in enumerate(self.weights_layer_4):
            fill_random[1,threads_per_block](self.weights_layer_4[i], 17, rng_states)


    def get_weights(self):
        w_layer2=np.reshape(self.weights_layer_2,(304))
        w_layer3=np.reshape(self.weights_layer_3,(272))
        w_layer4=np.reshape(self.weights_layer_4,(68))
        return np.concatenate((w_layer2,w_layer3,w_layer4),axis=None)
    
    def set_weights(self, weights):
        w_layer2=np.array(weights[:304])
        self.weights_layer_2=np.reshape(w_layer2,(16,19))
        
        w_layer3=np.array(weights[304:304+272])
        self.weights_layer_3=np.reshape(w_layer3,(16,17))
        
        w_layer4=np.array(weights[304+272:])
        self.weights_layer_4=np.reshape(w_layer4,(4,17))
      

    def predict(self, input_arr):  
        arr=np.append(input_arr,[[1.]],0)
        A_global_mem = cuda.to_device(self.weights_layer_2)
        B_global_mem = cuda.to_device(arr)
        C_global_mem = cuda.device_array((16, 1))
        threadsperblock = (16, 16)
        blockspergrid_x = int(math.ceil(self.weights_layer_2.shape[0] / threadsperblock[0]))
        blockspergrid_y = int(math.ceil(arr.shape[1] / threadsperblock[1]))
        blockspergrid = (blockspergrid_x, blockspergrid_y)
        # Start the kernel 
        matmul[blockspergrid, threadsperblock](A_global_mem, B_global_mem, C_global_mem) #calculated values of second layer
        C = C_global_mem.copy_to_host()
        for i in range(16):
            C[i][0]=expit(C[i][0])
        
        arr=np.append(C,[[1.]],0)
        A_global_mem = cuda.to_device(self.weights_layer_3)
        B_global_mem = cuda.to_device(arr)    
        C_global_mem = cuda.device_array((16, 1))
        blockspergrid_x = int(math.ceil(self.weights_layer_3.shape[0] / threadsperblock[0]))
        blockspergrid_y = int(math.ceil(arr.shape[1] / threadsperblock[1]))
        blockspergrid = (blockspergrid_x, blockspergrid_y)
        matmul[blockspergrid, threadsperblock](A_global_mem, B_global_mem, C_global_mem) #calculated values of third layer
   
        C = C_global_mem.copy_to_host()
        for i in range(8):
            C[i][0]=expit(C[i][0])
        
        arr=np.append(C,[[1.]],0)
        A_global_mem = cuda.to_device(self.weights_layer_4)
        B_global_mem = cuda.to_device(arr)
        C_global_mem = cuda.device_array((4, 1))
        blockspergrid_x = int(math.ceil(self.weights_layer_3.shape[0] / threadsperblock[0]))
        blockspergrid_y = int(math.ceil(arr.shape[1] / threadsperblock[1]))
        blockspergrid = (blockspergrid_x, blockspergrid_y)
        matmul[blockspergrid, threadsperblock](A_global_mem, B_global_mem, C_global_mem) #calculated values of the last layer
    
        # Copy the result back to the host
        C = C_global_mem.copy_to_host()
        C=np.reshape(C,(4))
        
        return np.where(C == np.amax(C))

    def check_direction(self, versor): 
        boolean = True
        body_found= False
        i=1
        distance_body=50
        distance_wall=50
        while boolean:
            pos=[x + i*y for x, y in zip(self.grid.snake.position[0], versor)]
            if pos[0]<0 or pos[0]>39 or pos[1]<0 or pos[1]>39:
                distance_wall = i
                boolean=False
            else:
                if self.grid.grid [pos[0]] [pos[1]] == 1:
                    if (not body_found):
                        distance_body = i 
                        body_found=True
            i+=1
        return [1/distance_body, 1/distance_wall]

    def calculate_distances(self):
        """calculate the distance(in x and y) between the head of the snake an the food. 
           Also looks in 8 directions to check if there is a part of the body of the snake or a wall"""
        self.grid.update_grid()
        distances=[]
        x_dist=self.grid.food.x-self.grid.snake.position[0][0]
        y_dist=self.grid.food.y-self.grid.snake.position[0][1]
        distances.append(x_dist)
        distances.append(y_dist)
        
        #versor (1,0)
        distances+=self.check_direction([1,0])
        #versor (1,1)
        distances+=self.check_direction([1,1])
        #versor (0,1)
        distances+=self.check_direction([0,1])
        #versor (-1,1)
        distances+=self.check_direction([-1,1])
        #versor (-1,0)
        distances+=self.check_direction([-1,0])
        #versor (-1,-1)
        distances+=self.check_direction([-1,-1])
        #versor (0,-1)
        distances+=self.check_direction([0,-1])
        #versor (1,-1)
        distances+=self.check_direction([1,-1])
        
        distances_np=np.array([distances])
        return distances_np
    
    def move_up(self):
        self.grid.move_up()
      
    def move_down(self):
        self.grid.move_down()
        
    def move_left(self):
        self.grid.move_left()

    def move_right(self):
        self.grid.move_right()

    def get_points(self):
        return self.grid.get_points()

    def get_food_pos(self):
        return self.grid.get_food_pos()

    def restart(self):
        self.grid.restart()
        self.moves_done=[]
        self.food_pos=[]
     
    #set of functions used only for replay (no random generationf of food)
    def Move_up(self):
        self.grid.Move_up()
      
    def Move_down(self):
        self.grid.Move_down()
        
    def Move_left(self):
        self.grid.Move_left()

    def Move_right(self):
        self.grid.Move_right()    
    

def get_points(Ia):
    return Ia.get_points()    
    
def get_score_2(ia):
    points=ia.grid.points
    moves_done=200+points*100-ia.grid.moves_left
    return (2**points)*moves_done*moves_done

def get_score_3(ia):
    points=ia.grid.points
    moves_done=200+points*100-ia.grid.moves_left
    return points+0.001*moves_done

def get_score(ia):
    points=ia.grid.points
    moves_done=200+points*100-ia.grid.moves_left

    if (points < 10):
      fitness = moves_done * moves_done * pow(2, points)
    else:
      fitness =  moves_done * moves_done
      fitness *= pow(2, 10)
      fitness *= points
    
    return fitness

def order_IA(ia):
    ia.sort(key=get_score, reverse=True)
    return ia

def play(ia):
    #0=up, 1=right, 2=down, 3=left

    while ia.grid.status:
        ia.food_pos.append(ia.get_food_pos())
        distances = ia.calculate_distances()
        distances = np.reshape(distances, (18,1))
        result = ia.predict(distances)
        
        if result[0][0]==0:
            ia.move_up()
            ia.moves_done.append(0)
        elif result[0][0]==1:
            ia.move_right()
            ia.moves_done.append(1)
        elif result[0][0]==2:
            ia.move_down()
            ia.moves_done.append(2)
        elif result[0][0]==3:
            ia.move_left()
            ia.moves_done.append(3)